File: src/main/java/earth/terrarium/tempad/common/network/messages/c2s/DeleteLocationPacket.java
package earth.terrarium.tempad.common.network.messages.c2s;

import com.teamresourceful.resourcefullib.common.network.Packet;
import com.teamresourceful.resourcefullib.common.network.base.PacketType;
import com.teamresourceful.resourcefullib.common.network.base.ServerboundPacketType;
import earth.terrarium.tempad.api.locations.LocationApi;
import earth.terrarium.tempad.common.Tempad;
import earth.terrarium.tempad.common.utils.TeleportUtils;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.entity.player.Player;

import java.util.UUID;
import java.util.function.Consumer;

public record DeleteLocationPacket(UUID location) implements Packet<DeleteLocationPacket> {

    public static final Handler HANDLER = new Handler();

    @Override
    public PacketType<DeleteLocationPacket> type() {
        return HANDLER;
    }

    public static class Handler implements ServerboundPacketType<DeleteLocationPacket> {
        public static final ResourceLocation ID = new ResourceLocation(Tempad.MODID, "delete_location");

        @Override
        public Class<DeleteLocationPacket> type() {
            return DeleteLocationPacket.class;
        }

        @Override
        public ResourceLocation id() {
            return ID;
        }

        @Override
        public DeleteLocationPacket decode(FriendlyByteBuf buf) {
            return new DeleteLocationPacket(buf.readUUID());
        }

        @Override
        public void encode(DeleteLocationPacket packet, FriendlyByteBuf buffer) {
            buffer.writeUUID(packet.location);
        }

        @Override
        public Consumer<Player> handle(DeleteLocationPacket message) {
            return player -> {
                if (TeleportUtils.hasTempad(player)) {
                    LocationApi.API.remove(player.level(), player.getUUID(), message.location);
                }
            };
        }
    }
}

File: src/main/kotlin/earth/terrarium/tempad/common/network/c2s/DeleteLocationPacket.kt
package earth.terrarium.tempad.common.network.c2s

import com.teamresourceful.bytecodecs.base.ByteCodec
import com.teamresourceful.bytecodecs.base.`object`.ObjectByteCodec
import com.teamresourceful.resourcefullib.common.bytecodecs.ExtraByteCodecs
import com.teamresourceful.resourcefullib.common.network.Packet
import com.teamresourceful.resourcefullib.common.network.base.NetworkHandle
import com.teamresourceful.resourcefullib.common.network.defaults.CodecPacketType
import earth.terrarium.tempad.tempadId
import earth.terrarium.tempad.api.locations.TempadLocations
import earth.terrarium.tempad.api.context.ContextHolder
import earth.terrarium.tempad.api.context.SyncableContext
import earth.terrarium.tempad.common.items.TempadItem
import earth.terrarium.tempad.common.registries.ModItems
import net.minecraft.resources.ResourceLocation
import java.util.UUID

data class DeleteLocationPacket(val ctx: ContextHolder<*>, val providerId: ResourceLocation, val locationId: UUID): Packet<DeleteLocationPacket> {
    companion object {
        val type = CodecPacketType.Server.create(
            "delete_location".tempadId,
            ObjectByteCodec.create(
                ContextHolder.codec.fieldOf(DeleteLocationPacket::ctx),
                ExtraByteCodecs.RESOURCE_LOCATION.fieldOf { it.providerId },
                ByteCodec.UUID.fieldOf { it.locationId },
                ::DeleteLocationPacket
            ),
            NetworkHandle.handle { message, player ->
                if ({ it.item === ModItems.tempad } !in player.inventory) return@handle
                val ctx = message.ctx.getCtx(player)
                TempadLocations[player, ctx, message.providerId]!! -= message.locationId
            }
        )
    }

    override fun type() = type
}
